h2o-algos/src/main/java/hex/tree/DTree.java

```diff
@@ -1364,6 +1461,424 @@ static Split findBestSplitPoint(DHistogram hs, int col, double min_rows, int con
+      valsUplift = MemoryManager.malloc8d(4*nbins);
+
+      for( int i=0; i<nbins; i++ ) {
+        int id = idxs[i];
+        vals[vals_dim*i+0] = hs._vals[vals_dim*id+0];
+        vals[vals_dim*i+1] = hs._vals[vals_dim*id+1];
+        vals[vals_dim*i+2] = hs._vals[vals_dim*id+2];
+        valsUplift[valsUpliftDim * i] = valsUplift[valsUpliftDim * id];
+        valsUplift[valsUpliftDim * i + 1] = valsUplift[valsUpliftDim * id + 1];
+        valsUplift[valsUpliftDim * i + 2] = valsUplift[valsUpliftDim * id + 2];
+        valsUplift[valsUpliftDim * i + 3] = valsUplift[valsUpliftDim * id + 3];
+        if (LOG.isTraceEnabled()) LOG.trace(vals[3*i] + " obs have avg response [" + i + "]=" + avgs[id]);
+      }
+    }
+
+    // Compute mean/var for cumulative bins from 0 to nbins inclusive.
+    double   wlo[] = MemoryManager.malloc8d(nbins+1);
+    double  wYlo[] = MemoryManager.malloc8d(nbins+1);
+    double wYYlo[] = MemoryManager.malloc8d(nbins+1);
+    double[]  numloTreat = MemoryManager.malloc8d(nbins + 1);
+    double[]  resploTreat = MemoryManager.malloc8d(nbins + 1);
+    double[] numloContr = MemoryManager.malloc8d(nbins + 1);
+    double[]  resploContr = MemoryManager.malloc8d(nbins + 1);
+
+    for( int b = 1; b <= nbins; b++ ) {
+      int id = vals_dim * (b - 1);
+      double n0 = wlo[b - 1], n1 = vals[id + 0];
+      if( n0==0 && n1==0 )
+        continue;
+      double m0 =  wYlo[b - 1], m1 = vals[id + 1];
+      double s0 = wYYlo[b - 1], s1 = vals[id + 2];
+      wlo[b] = n0+n1;
+      wYlo[b] = m0+m1;
+      wYYlo[b] = s0+s1;
+      id = valsUpliftDim * (b - 1);
+      double nt0 = numloTreat[b - 1], nt1 = valsUplift[id];
+      numloTreat[b] = nt0 + nt1;
+      double dt0 = resploTreat[b - 1], dt1 = valsUplift[id + 1];
+      resploTreat[b] = dt0 + dt1;
+      double nc0 = numloContr[b - 1], nc1 = valsUplift[id + 2];
+      numloContr[b] = nc0 + nc1;
+      double dc0 = resploContr[b - 1], dc1 = valsUplift[id + 3];
+      resploContr[b] = dc0 + dc1;
+    }
+    final double wNA = hs.wNA();
+    double tot = wlo[nbins] + wNA; //total number of (weighted) rows
+    // Is any split possible with at least min_obs?
+    if( tot < 2*min_rows ) {
+      if (LOG.isTraceEnabled()) LOG.trace("can't split " + hs._name + ": min_rows: total number of observations is " + tot);
+      return null;
+    }
+    // If we see zero variance, we must have a constant response in this
+    // column.  Normally this situation is cut out before we even try to split,
+    // but we might have NA's in THIS column...
+    double wYNA = hs.wYNA();
+    double wYYNA = hs.wYYNA();
+    double var = (wYYlo[nbins]+wYYNA)*tot - (wYlo[nbins]+wYNA)*(wYlo[nbins]+wYNA);
+    if( ((float)var) == 0f ) {
+      if (LOG.isTraceEnabled()) LOG.trace("can't split " + hs._name + ": var = 0.");
+      return null;
+    }
+
+    // Compute mean/var for cumulative bins from nbins to 0 inclusive.
+    double   whi[] = MemoryManager.malloc8d(nbins+1);
```

Comment:
`nbins+1` should also be in local variable when it is repeated multiple times

Reply:
Good point, but no refactor in this code is allowed within this PR.